File: src/llm_redactor/detect/llm_validator.py
"""LLM-based validation pass for detected spans.

After regex + NER detection, sends each span with surrounding context
to a local LLM and asks: "Is this actually sensitive information?"

This catches:
- False positives: drug names tagged as PERSON, abbreviations as ORG
- Missed entities: names in unusual formats that NER missed

The validator is opt-in (adds one Ollama round-trip per request) and
configurable via pipeline.llm_validation in the config.
"""

from __future__ import annotations

import json

import httpx

from .types import Span


# Context window: characters before/after the span to include.
_CONTEXT_CHARS = 80

_SYSTEM_PROMPT = """\
You are a PII/secret detection validator. You will be given a text span \
that was flagged as potentially sensitive, along with its surrounding context \
and the detection type.

Your job: determine if this span is ACTUALLY sensitive information that \
should be redacted before sending to a cloud LLM.

Rules:
- KEEP means "yes, redact this — it is real PII or a real secret"
- DROP means "no, this is a false positive — do not redact"
- Drug names, medical terms, technical jargon → DROP
- Common abbreviations (PII, API, SQL, Q3) when not actual identifiers → DROP
- Actual person names, emails, phone numbers, SSNs, API keys → KEEP
- Organization names that are real companies → KEEP
- Generic words (café, office, today) → DROP

Respond with ONLY a JSON object: {"verdict": "KEEP"} or {"verdict": "DROP"}
"""


async def validate_spans(
    text: str,
    spans: list[Span],
    *,
    endpoint: str = "http://127.0.0.1:11434",
    model: str = "llama3.2:3b",
    timeout: float = 30.0,
) -> list[Span]:
    """Validate detected spans using a local LLM.

    Returns only the spans that the LLM confirms as real sensitive content.
    Spans from regex detection (confidence=1.0) are auto-kept — only NER
    spans are validated to save LLM calls.
    """
    if not spans:
        return []

    # Regex detections are high-confidence by design — skip validation.
    regex_spans = [s for s in spans if s.source == "regex"]
    ner_spans = [s for s in spans if s.source != "regex"]

    if not ner_spans:
        return regex_spans

    # Batch all NER spans into one LLM call for efficiency.
    validated = list(regex_spans)
    kept = await _batch_validate(text, ner_spans, endpoint=endpoint, model=model, timeout=timeout)
    validated.extend(kept)

    return validated


async def _batch_validate(
    text: str,
    spans: list[Span],
    *,
    endpoint: str,
    model: str,
    timeout: float,
) -> list[Span]:
    """Validate a batch of spans in a single LLM call."""
    # Build the prompt with all spans and their context.
    entries = []
    for i, span in enumerate(spans):
        ctx_start = max(0, span.start - _CONTEXT_CHARS)
        ctx_end = min(len(text), span.end + _CONTEXT_CHARS)
        context = text[ctx_start:ctx_end]

        entries.append(
            f"Span {i+1}: \"{span.text}\" (detected as: {span.kind}, confidence: {span.confidence:.2f})\n"
            f"Context: ...{context}..."
        )

    prompt = (
        "Validate these detected spans. For each, respond KEEP or DROP.\n\n"
        + "\n\n".join(entries)
        + "\n\nRespond with ONLY a JSON array of verdicts, e.g.:\n"
        + '[{"span": 1, "verdict": "KEEP"}, {"span": 2, "verdict": "DROP"}]'
    )

    url = f"{endpoint.rstrip('/')}/api/chat"
    body = {
        "model": model,
        "messages": [
            {"role": "system", "content": _SYSTEM_PROMPT},
            {"role": "user", "content": prompt},
        ],
        "stream": False,
        "options": {"temperature": 0.0, "num_predict": 200},
    }

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.post(url, json=body)
            resp.raise_for_status()
            raw = resp.json().get("message", {}).get("content", "")
    except Exception:
        # If LLM is unreachable, keep all spans (fail safe).
        return list(spans)

    return _parse_verdicts(raw, spans)


def _parse_verdicts(raw: str, spans: list[Span]) -> list[Span]:
    """Parse LLM response and return kept spans."""
    # Try to extract JSON from the response.
    try:
        # Handle response wrapped in markdown code blocks.
        clean = raw.strip()
        if clean.startswith("```"):
            clean = clean.split("\n", 1)[1].rsplit("```", 1)[0].strip()

        verdicts = json.loads(clean)
        if isinstance(verdicts, list):
            drop_indices: set[int] = set()
            for v in verdicts:
                idx = v.get("span", 0) - 1  # 1-indexed
                if v.get("verdict", "").upper() == "DROP" and 0 <= idx < len(spans):
                    drop_indices.add(idx)
            return [s for i, s in enumerate(spans) if i not in drop_indices]
    except (json.JSONDecodeError, KeyError, TypeError):
        pass

    # Fallback: parse line-by-line for "KEEP" / "DROP".
    lines = raw.strip().splitlines()
    kept = []
    for i, span in enumerate(spans):
        if i < len(lines):
            if "DROP" in lines[i].upper():
                continue
        # Default: keep (fail safe).
        kept.append(span)

    return kept

File: burp-plugin/src/main/kotlin/com/llmredactor/burp/detect/LlmValidator.kt
package com.llmredactor.burp.detect

import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * Optional Ollama batch validation for NER spans (regex spans are kept as-is).
 * Port of src/llm_redactor/detect/llm_validator.py
 */
object LlmValidator {

    private const val CONTEXT_CHARS = 30

    private const val SYSTEM_PROMPT = """You are a PII/secret detection validator. You will be given a text span that was flagged as potentially sensitive, along with its surrounding context and the detection type.

Your job: determine if this span is ACTUALLY sensitive information that should be redacted before sending to a cloud LLM.

Rules:
- KEEP means "yes, redact this — it is real PII or a real secret"
- DROP means "no, this is a false positive — do not redact"
- Drug names, medical terms, technical jargon → DROP
- Common abbreviations (PII, API, SQL, Q3) when not actual identifiers → DROP
- Actual person names, emails, phone numbers, SSNs, API keys → KEEP
- Organization names that are real companies → KEEP
- Generic words (café, office, today) → DROP

Respond with ONLY a JSON object: {"verdict": "KEEP"} or {"verdict": "DROP"} for a single span, or a JSON array [{"span": 1, "verdict": "KEEP"}, ...] for multiple."""

    private val client: HttpClient = HttpClient.newBuilder()
        .version(HttpClient.Version.HTTP_1_1)
        .connectTimeout(Duration.ofSeconds(3))
        .build()

    fun validate(
        text: String,
        spans: List<Span>,
        endpoint: String,
        model: String,
        timeoutSeconds: Long = 30,
    ): List<Span> {
        if (spans.isEmpty()) return emptyList()

        val regexSpans = spans.filter { it.source == "regex" }
        val nerSpans = spans.filter { it.source != "regex" }
        if (nerSpans.isEmpty()) return regexSpans

        val deduped = mutableListOf<Span>()
        val keyToIndex = mutableMapOf<Pair<String, String>, Int>()
        for (s in nerSpans) {
            val key = s.text.trim().lowercase() to s.kind
            if (key !in keyToIndex) {
                keyToIndex[key] = deduped.size
                deduped.add(s)
            }
        }

        val keptUnique = batchValidate(text, deduped, endpoint, model, timeoutSeconds)
        val keptIndices = keptUnique.map { deduped.indexOf(it) }.toSet()

        val validated = regexSpans.toMutableList()
        for (s in nerSpans) {
            val idx = keyToIndex[s.text.trim().lowercase() to s.kind]
            if (idx != null && idx in keptIndices) validated.add(s)
        }
        return validated
    }

    private fun batchValidate(
        text: String,
        spans: List<Span>,
        endpoint: String,
        model: String,
        timeoutSeconds: Long,
    ): List<Span> {
        if (endpoint.isBlank() || model.isBlank()) return spans

        val entries = spans.mapIndexed { i, span ->
            val ctxStart = maxOf(0, span.start - CONTEXT_CHARS)
            val ctxEnd = minOf(text.length, span.end + CONTEXT_CHARS)
            val context = text.substring(ctxStart, ctxEnd)
            "Span ${i + 1}: \"${span.text}\" (detected as: ${span.kind}, confidence: ${"%.2f".format(span.confidence)})\n" +
                "Context: ...$context..."
        }

        val prompt = buildString {
            append("Validate these detected spans. For each, respond KEEP or DROP.\n\n")
            append(entries.joinToString("\n\n"))
            append("\n\nRespond with ONLY a JSON array of verdicts, e.g.:\n")
            append("[{\"span\": 1, \"verdict\": \"KEEP\"}, {\"span\": 2, \"verdict\": \"DROP\"}]")
        }

        val url = "${endpoint.trimEnd('/')}/api/chat"
        val body = JSONObject()
            .put("model", model)
            .put(
                "messages",
                JSONArray()
                    .put(JSONObject().put("role", "system").put("content", SYSTEM_PROMPT))
                    .put(JSONObject().put("role", "user").put("content", prompt)),
            )
            .put("stream", false)
            .put("options", JSONObject().put("temperature", 0.0).put("num_predict", 200))

        return try {
            val request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(timeoutSeconds))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
                .build()
            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() != 200) return spans
            val raw = JSONObject(response.body()).optJSONObject("message")?.optString("content").orEmpty()
            parseVerdicts(raw, spans)
        } catch (_: Exception) {
            spans
        }
    }

    private fun parseVerdicts(raw: String, spans: List<Span>): List<Span> {
        var clean = raw.trim()
        if (clean.startsWith("```")) {
            clean = clean.substringAfter('\n').substringBeforeLast("```").trim()
        }
        try {
            val verdicts = JSONArray(clean)
            val dropIndices = mutableSetOf<Int>()
            for (i in 0 until verdicts.length()) {
                val v = verdicts.get(i)
                when (v) {
                    is JSONObject -> {
                        val idx = v.optInt("span", i + 1) - 1
                        if (v.optString("verdict", "").uppercase() == "DROP" && idx in spans.indices) {
                            dropIndices.add(idx)
                        }
                    }
                }
            }
            return spans.filterIndexed { i, _ -> i !in dropIndices }
        } catch (_: Exception) {
            // line-by-line fallback
        }
        val lines = clean.lines()
        val kept = mutableListOf<Span>()
        for (i in spans.indices) {
            if (i < lines.size && "DROP" in lines[i].uppercase()) continue
            kept.add(spans[i])
        }
        return if (kept.isEmpty() && spans.isNotEmpty()) spans else kept
    }
}
